koku/api/provider/view.py

```diff
@@ -115,10 +118,13 @@ def get_queryset(self):
                 LOG.error('No customer found for user %s.', user)
         return queryset
 
+    @never_cache
     def create(self, request, *args, **kwargs):
         """Create a Provider."""
         return super().create(request=request, args=args, kwargs=kwargs)
 
+    @cache_control(private=True)
+    @vary_on_headers(RH_IDENTITY_HEADER)
     def list(self, request, *args, **kwargs):
```

Comment:
If we are going to be caching this, could we invalidate the cache whenever a provider is created or removed?

Reply:
We can, but as the saying goes, there are two Very Hard Problems in computing - naming things and cache invalidation.

To invalidate an object, we need to know it's key in the cache. The cache key is generated based on attributes and headers from the request and hashed values (i.e. the url is hashed and used as part of the key). The Vary header in our response specifies other headers to use as part of the cache key hash.

(See: https://github.com/django/django/blob/master/django/utils/cache.py#L304 )

If there's a concern about returning cached responses and there isn't a reasonable TTL we can set (e.g. would a 5-minute cache TTL work where a 60-minute TTL won't?), we probably shouldn't cache the view at all.

Reply:
Sure. I guess the thing that makes the most sense here is the caching around the reports view as that is where our real time is spent. Perhaps without expending a bunch of time on cache invalidation we just focus on caching just for reports and go with `@never_cache` for the rest? We could open an issue around the invalidation approach for other APIs.